jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -15,65 +15,113 @@
+ * <p>A {@code RetainableByteBuffer} can either be:
+ * <ul>
+ *     <li>in pool; in this case {@link #isRetained()} returns {@code false} and calling {@link #release()} throws {@link IllegalStateException}</li>
+ *     <li>out of pool but not retained; in this case {@link #isRetained()} returns {@code false} and calling {@link #release()} returns {@code true}</li>
+ *     <li>out of pool and retained; in this case {@link #isRetained()} returns {@code true} and calling {@link #release()} returns {@code false}</li>
+ * </ul>
+ * <p>Calling {@link #release()} on a out of pool and retained instance does not re-pool it while that re-pools it on a out of pool but not retained instance.</p>
  */
 public class RetainableByteBuffer implements Retainable
 {
-    private final ByteBufferPool pool;
     private final ByteBuffer buffer;
-    private final AtomicInteger references;
+    private final AtomicInteger references = new AtomicInteger();
+    private final Consumer<ByteBuffer> releaser;
+    private final AtomicLong lastUpdate = new AtomicLong(System.nanoTime());
 
-    public RetainableByteBuffer(ByteBufferPool pool, int size)
+    RetainableByteBuffer(ByteBuffer buffer, Consumer<ByteBuffer> releaser)
     {
-        this(pool, size, false);
+        this.releaser = releaser;
+        this.buffer = buffer;
     }
 
-    public RetainableByteBuffer(ByteBufferPool pool, int size, boolean direct)
+    public int capacity()
     {
-        this.pool = pool;
-        this.buffer = pool.acquire(size, direct);
-        this.references = new AtomicInteger(1);
+        return buffer.capacity();
     }
 
     public ByteBuffer getBuffer()
     {
         return buffer;
     }
 
-    public int getReferences()
+    public long getLastUpdate()
     {
-        return references.get();
+        return lastUpdate.getOpaque();
     }
 
+    /**
+     * Checks if {@link #retain()} has been called at least one more time than {@link #release()}.
+     * @return true if this buffer is retained, false otherwise.
+     */
+    public boolean isRetained()
+    {
+        return references.get() > 1;
+    }
+
+    public boolean isDirect()
+    {
+        return buffer.isDirect();
+    }
+
+    /**
+     * Increments the retained counter of this buffer. It must be done internally by
+     * the pool right after creation and after each un-pooling.
+     * The reason why this method exists on top of {@link #retain()} is to be able to
+     * have some safety checks that must know why the ref counter is being incremented.
+     */
+    void acquire()
+    {
+        if (references.getAndIncrement() != 0)
+        {
+            references.decrementAndGet();
+            throw new IllegalStateException("re-pooled while still used " + this);
+        }
+    }
+
+    /**
+     * Increments the retained counter of this buffer.
+     */
     @Override
     public void retain()
     {
-        while (true)
+        if (references.getAndIncrement() == 0)
```

Comment:
I think the following would avoid temporarily incrementing to 1, which other threads might see.
```suggestion
        if (references.getAndUpdate(c -> c == 0 ? 0 : c + 1) == 0)
```

Reply:
Good point, I'll do this change.